Sources/NIOHPACK/DynamicHeaderTable.swift

```diff
@@ -0,0 +1,169 @@
+    }
+    
+    // internal for testing
+    func dumpHeaders() -> String {
+        return self.storage.dumpHeaders(offsetBy: StaticHeaderTable.count)
+    }
+    
+    /// Searches the table for a matching header, optionally with a particular value. If
+    /// a match is found, returns the index of the item and an indication whether it contained
+    /// the matching value as well.
+    ///
+    /// Invariants: If `value` is `nil`, result `containsValue` is `false`.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header for which to search.
+    ///   - value: Optional value for the header to find. Default is `nil`.
+    /// - Returns: A tuple containing the matching index and, if a value was specified as a
+    ///            parameter, an indication whether that value was also found. Returns `nil`
+    ///            if no matching header name could be located.
+    func findExistingHeader<Name: Collection, Value: Collection>(named name: Name, value: Value?) -> (index: Int, containsValue: Bool)? where Name.Element == UInt8, Value.Element == UInt8 {
+        // looking for both name and value, but can settle for just name
+        // thus we'll search manually.
+        guard let value = value else {
+            return self.storage.indices(matching: name).first.map { ($0, false) }
+        }
+        
+        var firstNameMatch: Int? = nil
+        for index in self.storage.indices(matching: name) {
+            if firstNameMatch == nil {
+                // record the first (most recent) index with a matching header name,
+                // in case there's no value match.
+                firstNameMatch = index
+            }
+            
+            if self.storage.view(of: self.storage[index].value).matches(value) {
+                // this entry has both the name and the value we're seeking
+                return (index, true)
+            }
+        }
+        
+        // no value matches -- but did we find a name?
+        if let index = firstNameMatch {
+            return (index, false)
+        } else {
+            // no matches at all
+            return nil
+        }
+    }
+    
+    /// Appends a header to the table. Note that if this succeeds, the new item's index
+    /// is always zero.
+    ///
+    /// This call may result in an empty table, as per RFC 7541 § 4.4:
+    /// > "It is not an error to attempt to add an entry that is larger than the maximum size;
+    /// > an attempt to add an entry larger than the maximum size causes the table to be
+    /// > emptied of all existing entries and results in an empty table."
+    ///
+    /// - Parameters:
+    ///   - name: A collection of UTF-8 code points comprising the name of the header to insert.
+    ///   - value: A collection of UTF-8 code points comprising the value of the header to insert.
+    ///   - evictAutomatically: If `true`, the table will evict existing items to make room. Default is `true`.
```

Comment:
I still think that layers above this might find it useful when constraining the header table size below the protocol maximum. It can choose to catch an error and make a determination on whether to manually grow the buffer or to try again with eviction.

An alternative API might be to pass in a closure that acts as a delegate for the decision. That feels a little clunky though, when we're this far from Foundation.

Reply:
Hm, this is an internal method, so clients won't be interacting with it directly, and my anticipated use case could thus be served in another manner. Okay, I'll remove it. You win 😊